File: eyan_helper/src/main/java/eu/eyan/util/jgoodies/FormLayoutPlus.java
package eu.eyan.util.jgoodies;

import java.awt.Component;
import java.awt.Container;
import java.awt.Dimension;
import java.awt.LayoutManager2;

import com.jgoodies.forms.factories.CC;
import com.jgoodies.forms.layout.CellConstraints;
import com.jgoodies.forms.layout.ColumnSpec;
import com.jgoodies.forms.layout.FormLayout.LayoutInfo;
import com.jgoodies.forms.layout.RowSpec;

public class FormLayoutPlus implements LayoutManager2 {
	private final com.jgoodies.forms.layout.FormLayout formLayout;
	private int rows = 0;
	private final Container container;

	public FormLayoutPlus(Container container, String colSpec) {
		this.container = container;
		formLayout = new com.jgoodies.forms.layout.FormLayout(colSpec);
		container.setLayout(this);
	}

	public FormLayoutPlus(Container container, String colSpec, String rowSpec, Component component) {
		this(container, colSpec);
		appendRow(rowSpec, component);
	}

	public FormLayoutPlus(Container container, String colSpec, String rowSpec, FormLayoutPlus formLayout) {
		this(container, colSpec);
		appendRow(rowSpec, formLayout);
	}

	public int appendRow(String... rowSpecs) {
		for (String rowSpec : rowSpecs) {
			appendRow(RowSpec.decode(rowSpec));
			rows++;
		}
		return rows;
	}

	public Container getComponent() {
		return container;
	}
	
	public FormLayoutPlus appendRow(String rowSpec, Component component) {
		appendRow(RowSpec.decode(rowSpec));
		rows++;
		if(component!=null){
			container.add(component, CC.xy(1, rows));
		}
		return this;
	}

	public FormLayoutPlus appendRow(String rowSpec, FormLayoutPlus formLayout) {
		return appendRow(rowSpec, formLayout.getComponent());
	}
	
	/// DELEGATES ///
	public int hashCode() {
		return formLayout.hashCode();
	}

	public boolean equals(Object obj) {
		return formLayout.equals(obj);
	}

	public String toString() {
		return formLayout.toString();
	}

	public int getColumnCount() {
		return formLayout.getColumnCount();
	}

	public ColumnSpec getColumnSpec(int columnIndex) {
		return formLayout.getColumnSpec(columnIndex);
	}

	public void setColumnSpec(int columnIndex, ColumnSpec columnSpec) {
		formLayout.setColumnSpec(columnIndex, columnSpec);
	}

	public void appendColumn(ColumnSpec columnSpec) {
		formLayout.appendColumn(columnSpec);
	}

	public void insertColumn(int columnIndex, ColumnSpec columnSpec) {
		formLayout.insertColumn(columnIndex, columnSpec);
	}

	public void removeColumn(int columnIndex) {
		formLayout.removeColumn(columnIndex);
	}

	public int getRowCount() {
		return formLayout.getRowCount();
	}

	public RowSpec getRowSpec(int rowIndex) {
		return formLayout.getRowSpec(rowIndex);
	}

	public void setRowSpec(int rowIndex, RowSpec rowSpec) {
		formLayout.setRowSpec(rowIndex, rowSpec);
	}

	public void appendRow(RowSpec rowSpec) {
		formLayout.appendRow(rowSpec);
	}

	public void insertRow(int rowIndex, RowSpec rowSpec) {
		formLayout.insertRow(rowIndex, rowSpec);
	}

	public void removeRow(int rowIndex) {
		formLayout.removeRow(rowIndex);
	}

	public CellConstraints getConstraints(Component component) {
		return formLayout.getConstraints(component);
	}

	public void setConstraints(Component component, CellConstraints constraints) {
		formLayout.setConstraints(component, constraints);
	}

	public int[][] getColumnGroups() {
		return formLayout.getColumnGroups();
	}

	public void setColumnGroups(int[][] colGroupIndices) {
		formLayout.setColumnGroups(colGroupIndices);
	}

	public void addGroupedColumn(int columnIndex) {
		formLayout.addGroupedColumn(columnIndex);
	}

	public int[][] getRowGroups() {
		return formLayout.getRowGroups();
	}

	public void setRowGroups(int[][] rowGroupIndices) {
		formLayout.setRowGroups(rowGroupIndices);
	}

	public void addGroupedRow(int rowIndex) {
		formLayout.addGroupedRow(rowIndex);
	}

	public boolean getHonorsVisibility() {
		return formLayout.getHonorsVisibility();
	}

	public void setHonorsVisibility(boolean b) {
		formLayout.setHonorsVisibility(b);
	}

	public void setHonorsVisibility(Component component, Boolean b) {
		formLayout.setHonorsVisibility(component, b);
	}

	public void addLayoutComponent(String name, Component component) {
		formLayout.addLayoutComponent(name, component);
	}

	public void addLayoutComponent(Component comp, Object constraints) {
		formLayout.addLayoutComponent(comp, constraints);
	}

	public void removeLayoutComponent(Component comp) {
		formLayout.removeLayoutComponent(comp);
	}

	public Dimension minimumLayoutSize(Container parent) {
		return formLayout.minimumLayoutSize(parent);
	}

	public Dimension preferredLayoutSize(Container parent) {
		return formLayout.preferredLayoutSize(parent);
	}

	public Dimension maximumLayoutSize(Container target) {
		return formLayout.maximumLayoutSize(target);
	}

	public float getLayoutAlignmentX(Container parent) {
		return formLayout.getLayoutAlignmentX(parent);
	}

	public float getLayoutAlignmentY(Container parent) {
		return formLayout.getLayoutAlignmentY(parent);
	}

	public void invalidateLayout(Container target) {
		formLayout.invalidateLayout(target);
	}

	public void layoutContainer(Container parent) {
		formLayout.layoutContainer(parent);
	}

	public LayoutInfo getLayoutInfo(Container parent) {
		return formLayout.getLayoutInfo(parent);
	}
	////  END DELEGATES  ////
}

File: src/main/scala/eu/eyan/util/jgoodies/FormLayoutPlus.scala
package eu.eyan.util.jgoodies

import com.jgoodies.forms.layout.FormLayout
import java.awt.LayoutManager2
import com.jgoodies.forms.layout.ColumnSpec
import com.jgoodies.forms.layout.RowSpec
import com.jgoodies.forms.layout.CellConstraints
import java.awt.Component
import java.awt.Container
import com.jgoodies.forms.factories.CC

object FormLayoutPlus {
  def apply(container: Container, colSpec: String) = {
    val ret = new FormLayoutPlus(container, colSpec)
    container.setLayout(ret)
    ret
  }

  def apply(container: Container, colSpec: String, rowSpec: String, component: Component) = {
    val ret = new FormLayoutPlus(container, colSpec)
    container.setLayout(ret)
    ret.appendRow(rowSpec, component)
    ret
  }
}

class FormLayoutPlus private (container: Container, colSpec: String) extends LayoutManager2 {
  val formLayout = new FormLayout(colSpec)
  private var rows = 0

  override def hashCode() = formLayout.hashCode()

  override def equals(obj: Any): Boolean = formLayout.equals(obj)

  override def toString = formLayout.toString

  def getComponent = container

  def appendRow(rowSpecs: String*): Int = {
    for {rowSpec <- rowSpecs} { appendRow(RowSpec.decode(rowSpec)); rows += 1 }
    rows
  }

  def appendRow(rowSpec: String, component: Component): FormLayoutPlus = {
    appendRow(RowSpec.decode(rowSpec))
    rows += 1
    if (component != null) container.add(component, CC.xy(1, rows))
    this
  }

  def appendRow(rowSpec: String, formLayout: FormLayoutPlus): FormLayoutPlus = appendRow(rowSpec, formLayout.getComponent)

  /* DELEGATES */
  def getColumnCount = formLayout.getColumnCount

  def getColumnSpec(columnIndex: Int) = formLayout.getColumnSpec(columnIndex)

  def setColumnSpec(columnIndex: Int, columnSpec: ColumnSpec) = formLayout.setColumnSpec(columnIndex, columnSpec)

  def appendColumn(columnSpec: ColumnSpec) = formLayout.appendColumn(columnSpec)

  def insertColumn(columnIndex: Int, columnSpec: ColumnSpec) = formLayout.insertColumn(columnIndex, columnSpec)

  def removeColumn(columnIndex: Int) = formLayout.removeColumn(columnIndex)

  def getRowCount = formLayout.getRowCount

  def getRowSpec(rowIndex: Int) = formLayout.getRowSpec(rowIndex)

  def setRowSpec(rowIndex: Int, rowSpec: RowSpec) = formLayout.setRowSpec(rowIndex, rowSpec)

  def appendRow(rowSpec: RowSpec) = formLayout.appendRow(rowSpec)

  def insertRow(rowIndex: Int, rowSpec: RowSpec) = formLayout.insertRow(rowIndex, rowSpec)

  def removeRow(rowIndex: Int) = formLayout.removeRow(rowIndex)

  def getConstraints(component: Component) = formLayout.getConstraints(component)

  def setConstraints(component: Component, constraints: CellConstraints) = formLayout.setConstraints(component, constraints)

  def getColumnGroups = formLayout.getColumnGroups

  def setColumnGroups(colGroupIndices: Array[Array[Int]]) = formLayout.setColumnGroups(colGroupIndices)

  def addGroupedColumn(columnIndex: Int) = formLayout.addGroupedColumn(columnIndex)

  def getRowGroups = formLayout.getRowGroups

  def setRowGroups(rowGroupIndices: Array[Array[Int]]) = formLayout.setRowGroups(rowGroupIndices)

  def addGroupedRow(rowIndex: Int) = formLayout.addGroupedRow(rowIndex)

  def getHonorsVisibility = formLayout.getHonorsVisibility

  def setHonorsVisibility(b: Boolean) = formLayout.setHonorsVisibility(b)

  def setHonorsVisibility(component: Component, b: Boolean) = formLayout.setHonorsVisibility(component, b)

  def addLayoutComponent(name: String, component: Component) = formLayout.addLayoutComponent(name, component)

  def addLayoutComponent(comp: Component, constraints: Object) = formLayout.addLayoutComponent(comp, constraints)

  def removeLayoutComponent(comp: Component) = formLayout.removeLayoutComponent(comp)

  def minimumLayoutSize(parent: Container) = formLayout.minimumLayoutSize(parent)

  def preferredLayoutSize(parent: Container) = formLayout.preferredLayoutSize(parent)

  def maximumLayoutSize(target: Container) = formLayout.maximumLayoutSize(target)

  def getLayoutAlignmentX(parent: Container) = formLayout.getLayoutAlignmentX(parent)

  def getLayoutAlignmentY(parent: Container) = formLayout.getLayoutAlignmentY(parent)

  def invalidateLayout(target: Container) = formLayout.invalidateLayout(target)

  def layoutContainer(parent: Container) = formLayout.layoutContainer(parent)

  def getLayoutInfo(parent: Container) = formLayout.getLayoutInfo(parent)
  /* DELEGATES END */
}